### ISSUE TITLE
Setting -Dwarnings breaks feature detection in build.rs

### ISSUE THREAD

**Issue by piodul** [tokio-rs/tokio#5373]
I'm not sure whether this is considered a bug due to custom `RUSTFLAGS`, but treating warnings as errors in CI doesn't sound uncommon.

**Version**

`cargo tree | grep tokio`
```
test-tokio-fail v0.1.0 (/tmp/test-tokio-fail)
└── tokio v1.24.1
```

**Platform**

`Linux <cut> 6.0.17-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 4 16:00:03 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`

**Description**

Steps to reproduce:

```bash
rustup toolchain install 1.59.0
cargo new --bin test-tokio-fail
cd ./test-tokio-fail/
echo "tokio = \"=1.24.0\"" >> Cargo.toml # or 1.24.1, same issue

# We are treating warnings as errors in the CI
export RUSTFLAGS="-Dwarnings"
# Alternatively:
# export RUSTFLAGS="--cfg tokio_no_atomic_u64"

cargo +1.59.0 check
```

I get the following error:

```
   Compiling tokio v1.24.0
error[E0432]: unresolved import `crate::util::once_cell`
 --> /home/piodul/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.0/src/loom/std/atomic_u64_static_once_cell.rs:3:18
  |
3 | use crate::util::once_cell::OnceCell;
  |                  ^^^^^^^^^ could not find `once_cell` in `util`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `tokio` due to previous error
```

My short investigation led me to the tokio's build.rs file. It looks like `-Dwarnings` affects the autocfg probe [here](https://github.com/tokio-rs/tokio/blob/c390a62387fe7346951c8bc57ea2761614b83e82/tokio/build.rs#L101), seen when running `cargo check` with high verbosity (`-vv`):

```
     Running `/tmp/test-tokio-fail/target/debug/build/tokio-0b45242da01fd5fb/build-script-build`
[tokio 1.24.0] error: unused import: `std::sync::atomic::AtomicU64 as _`
[tokio 1.24.0]  --> <anon>:3:9
[tokio 1.24.0]   |
[tokio 1.24.0] 3 |     use std::sync::atomic::AtomicU64 as _;
[tokio 1.24.0]   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[tokio 1.24.0]   |
[tokio 1.24.0]   = note: `-D unused-imports` implied by `-D warnings`
[tokio 1.24.0] 
[tokio 1.24.0] error: aborting due to previous error
[tokio 1.24.0] 
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_target_has_atomic
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_const_mutex_new
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_atomic_u64
```

Because of this, `cargo:rustc-cfg=tokio_no_atomic_u64` is emitted in addition to others (I get the same effect with `RUSTFLAGS="--cfg tokio_no_atomic_u64"`). This cfg seems to be key to trigger the error.

**Comment by Darksonn** [comment]
Thank you for reporting this. Setting `-Dwarnings` shouldn't break the feature detection in our `build.rs` file.

The specific compilation failure you ran into was fixed in 1.24.1, but the problem persists — it will incorrectly detect features even if you upgrade, leading to less efficient code.

**Comment by taiki-e** [comment]
This is because once_cell is available only when rt, signal, or process is enabled, but atomic_u64_static_once_cell.rs is compiled unconditionally.

https://github.com/tokio-rs/tokio/blob/c390a62387fe7346951c8bc57ea2761614b83e82/tokio/src/util/mod.rs#L9-L10
https://github.com/tokio-rs/tokio/blob/c390a62387fe7346951c8bc57ea2761614b83e82/tokio/src/loom/std/mod.rs#L4-L5

Can be reproduced with:

```
RUSTFLAGS='--cfg tokio_no_atomic_u64 --cfg tokio_no_const_mutex_new' cargo hack build -p tokio --no-dev-deps
```
